Step3 - 로또(자동)

happy-dev-travel opened the pull request:
상세한 리뷰에 미리 감사드립니다!

happy-dev-travel committed:
feat(domain): 1차 객체 상호작용 설계, lotto 구입 기능 개발

LottoStore - lotto 구입 및 machine 작동
LottoMachine - lotto 번호 생성 담당
LottoBundle - lotto 꾸러미

happy-dev-travel committed:
feat(domain): LottoMachine의 발급 기능 추가

- 입력받은 갯수만큼 6개의 숫자를 가진 Lotto를 생성함

happy-dev-travel committed:
feat(domain): Lotto 번호가 정렬되도록 함

happy-dev-travel committed:
feat(domain): Lotto번호 출력 기능 추가

happy-dev-travel committed:
feat(domain): 맞춘 번호 갯수를 세는 기능 추가

happy-dev-travel committed:
feat(domain): 모든 lottoTicket의 정답 갯수를 세도록 추가

happy-dev-travel committed:
doc(documentation): README.md 작성

happy-dev-travel committed:
feat(domain): 당첨 갯수 와 수익률 계산 기능 추가

happy-dev-travel committed:
feat(controller): 프로그램 전체 로직 추가

happy-dev-travel committed:
feat(view): view 출력 추가

view가 출력 대상을 주입받아 유연하게 변경할 수 있도록 함.
view에 사용되는 메세지 포맷을 상수로 변경.

shared-moon commented on src/main/java/lotto/LottoApplication.java:
```diff
@@ -0,0 +1,28 @@
+import lotto.view.OutputTarget;
+
+public class LottoApplication {
+    public static final int lottoFee = 1000;
```
상수명은 보통 Upper snake case를 사용해요! LOTTO_FEE

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private Money getPaidMoney() {
+        this.view.showMessageRequestPurchaseMoney();
```
굳이 this를 명시해서 접근하지 않아도 괜찮지 않을까요 ?

happy-dev-travel replied:
네 맞습니다ㅎㅎ 저는 함수 인자랑 인스턴스 필드랑 이름이 같은 경우를 종종 봤어서 this를 붙이는게 약간 습관이 된 것 같네요ㅎㅎ 혹시 안 붙이는게 권장되는 사항인지요? 리뷰는 고쳐서 올리도록 하겠습니다! :)

shared-moon replied:
네 기본적으로 멤버변수와 같은 이름으로는 안만드는게 권장사항이라 정확히 말하면 this를 쓸 상황을 안만드는게 권장사항입니다! Setter나 생성자에서가 약간 예외 케이스? 라고 생각해주시면 좋을 것 같아요!

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,54 @@
+
+    private Money getPaidMoney() {
+        this.view.showMessageRequestPurchaseMoney();
+        return Money.of(Integer.parseInt(this.input.getInput()));
```
`parseInt()`는 숫자인지 확인하는 책임도 같이 가지고 있어서 `getInput()` 내부에서 int를 반환해주는 건 어떨까요 ?

happy-dev-travel replied:
말씀 감사합니다 :) 개발할 때는 input쪽 class를 최대한 단순하게 하고 싶었던 것 같습니다.  숫자만 받는 함수, 로또 정답을 받는 함수로 나눌 수도 있을 텐데 그렇게 하지 않았거든요. 그래서 숫자인지 확인하는걸 Money Class에 책임을 줄까 하다가 Controller에게 맡겼네요

shared-moon replied:
의도는 좋다고 생각합니다! 말씀드렸던 의도는 유효한 값에 대한 검증은 입력 받은 시점에 이루어지는게 좋지 않을까 하는 생각이었습니다 :)

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,54 @@
+
+    private List<Integer> getWinningNumbers() {
+        this.view.showMessageRequestWinningNumbers();
+        String input2 = this.input.getInput();
```
input2 말고 다른 명시적인 이름이 없을까요 ?

happy-dev-travel replied:
꼭 수정하겠습니다 👍

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,54 @@
+
+    private LottoBundle buyLotto(Money paidMoney) {
+        LottoBundle lottoBundle = this.lottoStore.buyLotto(paidMoney);
+        this.view.showLottoCount(lottoBundle.size());
```
io 작업이 비즈니스 로직 사이에 껴있는 것 같아요! `구매` 와 `출력` 을 구분하면 어떨까요 ?

happy-dev-travel replied:
Controller의 함수라서 크게 구매와 출력을 하나의 비즈니스 로직 함수로 묶어도 된다고 생각했던 것 같습니다. 출력을 하는건 다른 문제일 수 있겠네요 감사합니다

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,54 @@
+    private List<Integer> getWinningNumbers() {
+        this.view.showMessageRequestWinningNumbers();
+        String input2 = this.input.getInput();
+        return Arrays.stream(input2.split(",")).map(Integer::valueOf).collect(Collectors.toList());
```
입력받은 숫자가 6개가 아니어도 진행이 될 것 같아요!

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,38 @@
+    private final List<LottoNumber> numbers;
+
+    public Lotto(List<Integer> numbers) {
+        List<Integer> integers = Collections.unmodifiableList(numbers);
```
불변리스트로 변환은 왜 해주는걸까요 ?

happy-dev-travel replied:
일급컬렉션을 사용할 때, 외부에서 주입된 컬렉션이 외부에서 변경되면(요소 추가 삭제 ) 일급 컬렉션에 영향을 줄 수 있어서 위와 같이 처리를 해준다고 들어서 적용해 봤습니다 :)

shared-moon replied:
객체 내부의 값이 외부와 격리가 되어야 한다면, 새로운 컬렉션을 만들어주기만 해도 주입받은 곳과는 연결이 끊어질 것 같아요! 말씀하신 부분은 내부의 값을 외부로 전달하는 경우에 신경을 쓰면 더 좋을 것 같네요 :)

그리고 여기서 말씀드렸던 이유는 사실 불변리스트로 만들자마자 바로 밑에서 stream() 연산을 통해 새로운 리스트로 만들어내고 있어서 .. 의미가 있나 싶어 여쭤봤어요!

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,38 @@
+                .map(LottoNumber::of)
+                .collect(Collectors.toList());
+
+        Collections.sort(this.numbers);
```
정렬이 필요하다면, 위에 stream() 작업에 중간연산으로 들어가는 건 어떨까요 ?

happy-dev-travel replied:
감사합니다 미션 안내에 정렬은 Collections.sort로 할 수 있다고 본 바람에 굳이 저렇게 작성했네요ㅎㅎ

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,38 @@
+        return Collections.unmodifiableList(this.numbers);
+    }
+
+    public String print(){
```
출력을 담당하는 기분이 드는 메서드 명이네요! toString() 오버라이드를 활용해보는 건 어떨까요 ?

happy-dev-travel replied:
감사합니다!

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,38 @@
+        return String.format("[%s]",convertToStringNumbers());
+    }
+
+    private String convertToStringNumbers() {
```
메서드를 이렇게 분리 한 이유가 있을까요 ?

happy-dev-travel replied:
지금 생각하면 큰 의미는 없는 듯 한데, 로또 번호를 출력하는 형식이 대괄호에서 다른 것으로 변경될 수 있다는데 초점을 뒀었습니다. 그래서 숫자를 join하는 함수와 대괄호에 넣는 함수가 분리되었구요. 사실 view 단에서 처리해야 할까 하다가 결정을 못해었습니다

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,38 @@
+                .collect(Collectors.joining(","));
+    }
+
+    public int check(List<Integer> winningNumbers) {
```
당첨번호의 묶음도 하나의 Lotto 객체로 바라보면, Lotto의 비교 결과를 확인한다는 뜻의 메서드 네이밍도 가능할 것 같아요!

happy-dev-travel replied:
아.. 당첨번호도 Lotto로...! 감사합니다 👍 👍 👍

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,38 @@
+
+    public int check(List<Integer> winningNumbers) {
+        return (int)winningNumbers.stream()
+                .map(w -> numbers.contains(LottoNumber.of(w)))
```
이건 그냥 filter만 써도 될 것 같은데요 ?

happy-dev-travel replied:
그런가요? 현재로는 winningNumber는 Integer이고 Lotto내부의 숫자들은 LottoNumber라는 객체로 wrapping되어 있어서요
filter만 쓰는 방향으로 가보겠습니다 :)

shared-moon replied:
오 네 map을 filter로만 바꿔도 똑같은 결과가 나올 것 같아요 :) 어차피 map에서 하는 연산이 boolean을 리턴하는 과정이라 ..

shared-moon commented on src/main/java/lotto/domain/LottoBundle.java:
```diff
@@ -0,0 +1,35 @@
+    private final List<Lotto> lottoTickets;
+
+    public LottoBundle(List<Lotto> lottoTickets) {
+        this.lottoTickets = Collections.unmodifiableList(lottoTickets);
```
개인적으로 클래스 내부에 세팅되는 변수를 불변으로 치환할 필요는 없을 것 같아요. 변경 할 일이 있더라도 private이라 클래스 내부에서 통제가 가능 할 거고, 불변이 아닌 경우 일어나는 문제는 대부분 바깥으로 객체를 열어주거나 같은 레퍼런스를 전달해 줌으로써 일어나는 것이기 때문에 굳이 필요가 없어 보입니다!

happy-dev-travel replied:
네 말씀 감사합니다 :)

shared-moon commented on src/main/java/lotto/domain/LottoBundle.java:
```diff
@@ -0,0 +1,35 @@
+        this.lottoTickets = Collections.unmodifiableList(lottoTickets);
+    }
+
+    public List<String> printAll() {
```
여기도 toString() 오버라이드로 해결하면 좋을 것 같아요!

happy-dev-travel replied:
그렇네요ㅎㅎ 그동안 toString()을 너무 안 썼다는 생각이 듭니다

shared-moon commented on src/main/java/lotto/domain/LottoBundle.java:
```diff
@@ -0,0 +1,35 @@
+        return lottoTickets.size();
+    }
+
+    private List<Integer> checkAll(List<Integer> winningNumber) {
```
보통 private은 public 밑에 위치시켜요!

happy-dev-travel replied:
평소 잘 생각하지 못한 부분이네요 감사합니다!!

shared-moon commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,25 @@
+        return lottoTickets;
+    }
+
+    public Lotto issue() {
```
issue는 혹시 무슨 의미로 쓰인걸까요 ?

happy-dev-travel replied:
Lotto 하나를 발행(생성) 한다의 의미였습니다ㅎㅎ 의미 전달이 바로 안 되었군요ㅠ

shared-moon replied:
이건 약간 취향 강요일 수 있어서 조심스러운데요, 메서드 명을 정할때는 최대한 보편적인 단어를 우선하는게 좋다고 생각해요! 보편적인 기준은 기본 라이브러리나 주로 쓰는 프레임워크 등에서 사용하는 표현 혹은 일상적인 단어가 될 것 같은데요, 예를 들면 메인 기능이 생성이라는 의미라면 create~ 조회라면 find~ 요청이라면 fetch~ 등등 .. 더 자주 쓰는 표현을 위주로 생각해보시면 더 좋을 것 같아요! (제가 영어가 짧아서 그럴수도 있습니다 ㅎㅎ)

shared-moon commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public static LottoNumber of(Integer number) {
+        if(Objects.isNull(number) || number <= 0 || number > 45){
```
범위는 상수로 빼보면 어떨까요 ?

happy-dev-travel replied:
네 알겠습니다!

shared-moon commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.Objects;
+
+public class LottoNumber implements Comparable<LottoNumber>{
```
VO 활용 좋습니다 👍

shared-moon commented on src/main/java/lotto/domain/LottoStore.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+public class LottoStore {
+    private final int lottoFee;
```
로또 가격은 상수로 여기에 정의되어도 괜찮을 것 같은데 어떻게 생각하시나요 ?

happy-dev-travel replied:
네네 맞습니다 작성할 때는 로또 한장 가격을 편하게 변경할 수 있게 해야지 라는 생각으로 생성자에서 주입받도록 했었습니다ㅎㅎ

shared-moon commented on src/main/java/lotto/domain/WinningBonus.java:
```diff
@@ -0,0 +1,34 @@
+    WinningBonus(int matchCount, int bonus) {
+
+        this.matchCount = matchCount;
+        this.bonus = bonus;
```
보너스의 의미가.. 당첨금 인가요 ?

happy-dev-travel replied:
네네 맞습니다!

shared-moon commented on src/main/java/lotto/domain/WinningMoney.java:
```diff
@@ -0,0 +1,37 @@
+        return money.divide(totalBonus);
+    }
+
+    private Integer calcBonusFromTable(List<WinningBonus> bonuses, Integer c) {
```
파라미터나 람다식에서나 c 같은 단순 표현은 지양해주세요. 길어지더라도 최대한 의미가 드러나게 쓰는 게 좋을 것 같습니다!

happy-dev-travel replied:
감사합니다! 평소에 계속 단순 표현으로 사용하고 있었네요ㅠㅠ

shared-moon reviewed (changes requested):
동길님 안녕하세요!
3단계 미션 잘 진행 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데, 확인 후 다시 리뷰 요청 부탁드릴게요!
화이팅입니다 🔥

happy-dev-travel reviewed (commented):
많은 배움이 되었습니다 반영해서 다시 리뷰 요청드리겠습니다 :)

happy-dev-travel committed:
refactor(controller): controller 리뷰 사항 반영

1. 불필요한 this 접근자 제거
2. Input값의 Parsing책임을 Controller -> LottoUserInput로 옮김
3. buyLotto 메소드가 가진 출력 로직을 분리
4. naming 변경
5. 당첨번호 갯수가 6개일 때만 진행될 수 있도록 변경

happy-dev-travel committed:
refactor(domain): 당첨번호 리스트를 Lotto로 담아 처리

1. winningNumbers -> Lotto로 처리하도록 변경
2. Lotto의 번호 정렬을 stream에서 처리
3. 출력을 위한 문자열 생성은 toString() 활용
4. 매직넘버 변경

happy-dev-travel committed:
refactor(domain): 구매가능 Lotto갯수를 Money class로 캡슐화

happy-dev-travel committed:
refactor(domain): Lotto가 당첨 Lotto와 비교하여 Rank를 반환하도록 변경

shared-moon commented on src/main/java/lotto/controller/ConsoleLottoInput.java:
```diff
@@ -1,8 +1,12 @@
+import java.util.stream.Collectors;
 
 public class ConsoleLottoInput implements LottoUserInput {
+    private static final int LOTTO_NUMBER_COUNT = 6;
```
👍

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -8,31 +8,33 @@ public class Lotto {
-    private String convertToStringNumbers() {
-        return this.numbers.stream()
+    @Override
+    public String toString() {
```
👍

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -8,31 +8,33 @@ public class Lotto {
     public Lotto(List<Integer> numbers) {
-        List<Integer> integers = Collections.unmodifiableList(numbers);
-        this.numbers = integers.stream()
+        this.numbers = numbers.stream()
```
👍

shared-moon commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -8,31 +8,33 @@ public class Lotto {
+        return String.format("[%s]", joinedLottoNumbers);
+    }
+
+    public Rank getRankBy(Lotto winningLotto) {
```
👍

shared-moon commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Optional;
+
+public enum Rank {
+    FOURTH(3,5000),
```
숫자는 1000단위로 `_` 로 나눠주면 더 읽기 좋을 것 같아요!

shared-moon reviewed (approved):
동길님 안녕하세요!
피드백 반영 잘 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데, 확인 후 다음 미션 진행 해 주세요!
고생 많으셨습니다 :)
